examples-next/auth/schema.ts

```diff
@@ -0,0 +1,144 @@
+import { createSchema, list } from '@keystone-spike/keystone/schema';
+import { text, checkbox, password } from '@keystone-spike/fields';
+
+/**
+ * TODO: with the new session API, we're breaking keystone's current assumptions about the
+ * `authentication` argument provided to access control functions. This will be a breaking change
+ * for developers using Keystone.
+ *
+ * This needs to be implemented (replace `authentication` with `session`) before the code below
+ * will work.
+ *
+ * While we're here I'd like to propose some other changes to improve the ergonomics; see the types
+ * below for my proposal. Note that I'm specifically omitting the graphql context and any way of
+ * executing queries because I'd like to review that API as well, separately (and we can add it in
+ * later). This may be a bit contentious, but we need a new object anyway so I'm happy if in the
+ * first pass we just implement ({ session, item }) argument so the code below works and we can
+ * discuss the rest of the proposal as a group.
+ *
+ * Note that custom schema access control deviates significantly from the list and field access
+ * control arguments; we need to revisit that in the context of how graphql schema extensions work
+ * now. See https://www.keystonejs.com/api/access-control#custom-schema-access-control
+ *
+ * Also note, I'm not sure the reasoning behind why ListAccess seems to be resolved once per
+ * mutation for mutations that operate on multiple items, while FieldAccess seems to be resolved
+ * once per item (many times per mutation) for the same. At least, that's what I think is going on.
+ *
+ * This seems like a real foot-gun, and I'd prefer that they're implemented consistently, in favor
+ * of aligning on FieldAccessInput and resolving the function once per item being updated in the
+ * case of mutations that operate on multiple items.
```

Comment:
> _I'm not sure the reasoning behind why ListAccess seems to be resolved once per mutation for mutations that operate on multiple items, while FieldAccess seems to be resolved once per item_

Because there's only one list, but there are multiple items. No need to re-resolve access on the list level over and over again when the result will be the same every time.

Reply:
That "resolve the list level" thing makes sense for create, but this is more about "item" and "field" level access -- if you're looking at the item data at all for an update check, with multiple itemIds, how do you say "some of these but not others"? does the entire operation fail? how do we return which items you _don't_ have access to? There may be answers but they're not in our docs.

And I would imagine it leads to more complex access control functions because you always need to code for _either_ single or _multiple_ `itemId(s)` being provided and handle each case correctly.

I think it's simpler and _much_ safer to only write these functions to check a single item... and call it one time each for multi-item mutations. (and by safer, access control functions are really important for developers to get exactly right, so the bar for "it works how you think it works" is really high here)

Arguably this is less optimised when you don't need to access item data to resolve the access, but then again if you're checking anything on the item data, you have to resolve it for each item anyway, and I'd go for safety over optimisation here

Reply:
Yes, I think we're talking about the same thing 👍 

On a `updateMany` style mutation, it will check list-level access once, and will check field-level access once per item to be updated.